mcs/class/Microsoft.Build.Utilities/Microsoft.Build.Utilities/ProcessWrapper.cs

```diff
@@ -74,6 +74,7 @@ public new void Start ()
 		public void WaitForOutput (int milliseconds)
 		{
 			CheckDisposed ();
+			WaitForExit (milliseconds);
 			WaitHandle.WaitAll (new WaitHandle[] { endEventOut, endEventErr, endEventExit }, milliseconds);
```

Comment:
@ludovic-henry I tried your suggestion, but I still hit the same issue! I did check that `WaitForExit` was indeed called in the `if` block, but still when `ExitCode` was accessed after that, it threw the same exception!

```
/Users/ankit/dev/xamarin-android/build-tools/mono-runtimes/mono-runtimes.targets: error : Error executing task Exec: Process must exit before requested information can be determined.
                                Error executing task Exec: System.InvalidOperationException: Process must exit before requested information can be determined.
                                  at System.Diagnostics.Process.EnsureState (System.Diagnostics.Process+State state) [0x000b9] in /Users/ankit/dev/mono/mcs/class/referencesource/System/services/monitoring/system/diagnosticts/Process.cs:1439
                                  at System.Diagnostics.Process.get_ExitCode () [0x00000] in /Users/ankit/dev/mono/mcs/class/referencesource/System/services/monitoring/system/diagnosticts/Process.cs:219
                                  at (wrapper remoting-invoke-with-check) System.Diagnostics.Process:get_ExitCode ()
                                  at Microsoft.Build.Utilities.ToolTask.ExecuteTool (System.String pathToTool, System.String responseFileCommands, System.String commandLineCommands) [0x00101] in /Users/ankit/dev/mono/mcs/class/Microsoft.Build.Utilities/Microsoft.Build.Utilities/ToolTask.cs:185
                                  at Microsoft.Build.Tasks.Exec.ExecuteTool (System.String pathToTool, System.String responseFileCommands, System.String commandLineCommands) [0x00026] in /Users/ankit/dev/mono/mcs/class/Microsoft.Build.Tasks/Microsoft.Build.Tasks/Exec.cs:83
                                  at Microsoft.Build.Utilities.ToolTask.Execute () [0x0001c] in /Users/ankit/dev/mono/mcs/class/Microsoft.Build.Utilities/Microsoft.Build.Utilities/ToolTask.cs:128
                                  at Microsoft.Build.BuildEngine.TaskEngine.Execute () [0x00000] in /Users/ankit/dev/mono/mcs/class/Microsoft.Build.Engine/Microsoft.Build.BuildEngine/TaskEngine.cs:134
                                  at Microsoft.Build.BuildEngine.BuildTask.Execute () [0x0008f] in /Users/ankit/dev/mono/mcs/class/Microsoft.Build.Engine/Microsoft.Build.BuildEngine/BuildTask.cs:101
```
And then
```
Unhandled Exception:
System.ObjectDisposedException: Cannot access a disposed object.
Object name: 'handle'.
  at System.Runtime.InteropServices.SafeHandle.DangerousAddRef (System.Boolean& success) [0x00025] in /Users/ankit/dev/mono/mcs/class/corlib/System.Runtime.InteropServices/SafeHandle.cs:125
  at System.Threading.NativeEventCalls.SetEvent (Microsoft.Win32.SafeHandles.SafeWaitHandle handle) [0x00002] in /Users/ankit/dev/mono/mcs/class/corlib/System.Threading/NativeEventCalls.cs:54
  at System.Threading.EventWaitHandle.Set () [0x00012] in /Users/ankit/dev/mono/mcs/class/corlib/System.Threading/EventWaitHandle.cs:207
  at (wrapper remoting-invoke-with-check) System.Threading.EventWaitHandle:Set ()
  at Microsoft.Build.Utilities.ProcessWrapper.Dispose (System.Boolean disposing) [0x0001d] in /Users/ankit/dev/mono/mcs/class/Microsoft.Build.Utilities/Microsoft.Build.Utilities/ProcessWrapper.cs:100
  at System.ComponentModel.Component.Finalize () [0x00000] in /Users/ankit/dev/mono/mcs/class/referencesource/System/compmod/system/componentmodel/Component.cs:35
```

And I agree with @grendello. The "expectation" from the `WaitForOutput` method is (this is what the code was earlier) that it returns when the process is done and we can use `ExitCode`. So, we should ensure that the process has indeed exited, so, always call `WaitForExit`.